File: src/utils/events.ts
import { listen } from "@tauri-apps/api/event";
import { AudioSession } from "../types/audioSession";

export enum AppEvent {
  VolumeChange = "volume-change-event",
  MixerVisibilityChange = "mixer-visibility-change-event",
  ThemeChange = "theme-change-event",
}

export interface EventPayloads {
  [AppEvent.VolumeChange]: AudioSession;
  [AppEvent.MixerVisibilityChange]: boolean;
  [AppEvent.ThemeChange]: string;
}

export function listenToEvent<T extends AppEvent>(event: T, callback: (payload: EventPayloads[T]) => void): Promise<() => void> {
  return listen(event, (event) => {
    callback(event.payload as EventPayloads[T]);
  });
}

File: src-tauri/src/events.rs
#[derive(Clone, Copy)]
pub enum AppEvent {
    VolumeChange,
    MixerVisibilityChange,
}

impl AppEvent {
    pub fn as_str(&self) -> &'static str {
        match self {
            AppEvent::VolumeChange => "volume-change-event",
            AppEvent::MixerVisibilityChange => "mixer-visibility-change-event",
        }
    }
}
